crates/fiber/src/lib.rs

```diff
@@ -1,13 +1,32 @@
+#![no_std]
+
+// no_std support: we can only support `panic=abort` builds, because the
+// `core::panic` module otherwise does not provide us the catch/resume
+// unwinding primitives we need to propagate unwinding across stacks.
+#[cfg(all(panic = "unwind", not(feature = "std")))]
+compile_error!(
+    "no_std builds with wasmtime-fiber (e.g., Wasmtime async) only support `panic=abort`."
+);
```

Comment:
Personally I think it's reasonable to skip the `compile_error!` here and just let this happen as normal. With `panic = "abort"` nothing should be unsafe, it just means the extra unwind bits here aren't too useful (and they'll get optimized out)

Reply:
This error only happens when using `panic = "unwind"`, which is unsound if not using `catch_unwind`.

Reply:
Ah yeah sorry that's just "faking" a non-inlined call to something that can theoretically unwind. You don't actually need `C-unwind` but with `panic=unwind` all normal rust functions can possibly unwind so a catch-and-panic should be inserted for all calls. Basically I'm still not sure how `#![no_std]` + `-Cpanic=unwind` is unsound myself

Reply:
I see: my confusion is the dual to your confusion I think, that is, I'm still not sure how `#![no_std]` + `-Cpanic=unwind` is sound :-)

Restating mostly for my own benefit, but to verify we're on the same page: so with the current state of `main`, we wrap a call to the entrypoint on the new stack in a call to `catch_unwind` (e.g. [here](https://github.com/bytecodealliance/wasmtime/blob/db956669a5bc7de1ac247c63b4ad895965469144/crates/fiber/src/lib.rs#L225)). If the fiber's body panics while on the fiber stack, system `libunwind` reaches this point, the panic is reified as a `RunResult`, and that `RunResult` is passed back to the caller on the original stack and rethrown with `resume_unwind`.

With this PR, in `no_std`, we don't have `catch_unwind`/`resume_unwind`. So a panic happens in a fiber's body; say that there is a custom unwinder (there must be I think, for one to be doing unwinding at all in `no_std`?); this unwinder iterates up the stack frames on the fiber stack. It reaches the top, there's no callsite in a `catch_unwind`; this is an unwind error / UB / something bad. Or is this prevented some other way?

Reply:
Oh sorry! Good point getting on the same page (can also talk more during the Cranelift meeting tomorrow if helpful)

I think the main thing to add is that `extern "C" fn foo() { panic!() }` unconditionally, and soundly, aborts the process. That happens irrespective of `-Cunwind=...` settings. Starting in Rust 1.81 all `extern "C"` functions have an implicit catch-and-abort. That means that even if we were to remove the `catch_unwind` happening in wasmtime-fiber the crate would still be sound, just not too helpful on panics. The `catch_unwind` is mostly still there to assist with debugging panics by shepherding the panic across the boundary. That's always optional though since it's sound to abort the process at any time (just better to not do so if possible).

That to me is why it's currently sound if we skip the `catch_unwind`. A bit of a historical note though is that Rust 1.81 is very recent and prior versions of rustc didn't catch-and-abort at `extern "C"` boundaries. That means that historically `catch_unwind` was indeed required for soundness but that's no longer the case.

Reply:
Ah! OK, that's the detail I had missed; thanks for your patience here. I guess since our MSRV is now 1.81, we can rely on this behavior unconditionally. I'll remove the compile-error and add a comment detailing this.